uvicorn/protocols/http/httptools_impl.py

```diff
@@ -222,10 +214,20 @@ def on_header(self, name: bytes, value: bytes):
 
     def on_headers_complete(self):
         http_version = self.parser.get_http_version()
+        method = self.parser.get_method()
+        self.scope["method"] = method.decode("ascii")
         if http_version != "1.1":
             self.scope["http_version"] = http_version
         if self.parser.should_upgrade():
             return
```

Comment:
Does it matter that `"path"`, `"raw_path"`, `"query_string"` are now not populated in the upgrade case?

Reply:
I'd be tempted to say no since the `handle_upgrade` does not need them and the ws protocols will build the scope themselves down the road.
this said it doesn't hurt I think to put it before, both ways pass the tests, would you prefer it that way ?

Reply:
let me know if this is ok for you this way @tomchristie 
I added latest master changes

Reply:
I think I'm okay with this, yup. 👍

An alternate approach would be to keep the change footprint absolutely as low as possible. I'm almost always in favour of the lowest possible change footprint in PRs, because they're easier to review and lower risk.

In this case we *could alternately* approach the PR like this...

```python
    def on_message_begin(self):
         self.url = b""

    def on_url(self, url):
         self.url += url

    def on_url_complete(self):
        # This isn't an `httptools` callback, but we need it because `on_url` can actually
        # be called multiple times, and we don't know on each call if it's complete or not.
        # Instead into this method from `on_headers_complete`, so that we've got a single
        # point at which the URL is set.
        ...  # Existing body of `on_url()`

    def on_headers_complete(self):
        self.on_url_complete()
        ...  # Existing body of `on_headers_complete()`
```

Which would result in a really small changeset. Which as I say, I tend to think is a great thing.

Having said that, it's not a *super* complex PR. It looks good already, and I don't want to give you extra work, so gonna okay this and then leave the final decision to you.

Reply:
ok will merge that way, I dont have the time !